File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack stk = new Stack();
        for(char ch: s.toCharArray()){
            if(ch == '(' || ch == '{' || ch == '[')
                stk.push(ch);
            else{
                if(stk.size()==0)
                    return false;
                if(Math.abs((char)stk.pop()-ch) > 3)
                    return false;
            }
        }
        return stk.size()==0;
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.rb
# @param {String} s
# @return {Boolean}
def is_valid(s)
  bracket_hash = { ')' => '(', '}' => '{', ']' => '[' }
  stack = []
  s.chars.each do |char|
      if bracket_hash.values.include?(char)
          stack.push(char)
      elsif bracket_hash.keys.include?(char)
          if stack.empty?
              return false
          elsif stack.pop != bracket_hash[char]
              return false
          end
      end
  end

  stack.empty?
end